File: 0322-coin-change/0322-coin-change.java
class Solution {
     int help(int idx , int coins[], int amount, int[][] dp){
         if(amount ==0){
            return 0;
        }
         if(idx == coins.length && amount != 0){
           
            return Integer.MAX_VALUE-1;
        }
        
       
        
        if(dp[idx][amount] != -1) return dp[idx][amount];
        
        int pick = Integer.MAX_VALUE, npick = Integer.MAX_VALUE;
        
        if(coins[idx] <= amount){
            pick = 1 + help(idx,coins,amount-coins[idx],dp);
        }
        npick = help(idx+1,coins,amount,dp);
        
        return dp[idx][amount] = Math.min(pick,npick);
    }
    public int coinChange(int[] coins, int amount) {
         int[][] dp = new int[coins.length][amount+1];
        for(int[] a: dp){
            Arrays.fill(a,-1);
        }
        int ans = help(0,coins, amount,dp);
        if(ans == Integer.MAX_VALUE-1) return -1;
        return ans;
    }
}

File: 0322-coin-change/0322-coin-change.cpp
class Solution {
public:
   int helper(int idx,int amount,vector<int>&coins,vector<vector<int>>&dp){
    
    if(idx>=coins.size())return 1e7;
    if(amount == 0) return 0;
    

    if(dp[idx][amount] != -1) return dp[idx][amount];

    int pick = INT_MAX;
    if(coins[idx]<=amount){
        
        pick = 1 + helper(idx,amount - coins[idx],coins,dp);
    }

    int nopick = helper(idx+1,amount,coins,dp);

    return dp[idx][amount] = min(pick,nopick);
   }
    int coinChange(vector<int>& coins, int amount) {
        
        vector<vector<int>>dp(coins.size()+1,vector<int>(amount+1,-1));
       int cnt = 0;
       int ans = helper(0,amount,coins,dp);

       if(ans == 1e7) return -1;
       return ans;

    }
};